ADBDEV-4149 Support Explicit Redistribute Motion inside the SubPlans.

bimboterminator1 opened the pull request:
Flow structure has a segidColIdx field, which referencing a gp_segment_id
column in plan's targetList when the Explicit Redistribute Motion is requested.
There was a problem, that _copyFlow, _equalFlow, _outFlow and _readFlow
functions did not handle the segidColIdx field. Therefore, the Flow node
could not be serialized and deserialized, or copied correctly.

The problem manifested itself when a query had UPDATE/DELETE operation, that
required Explicit Redistribute, inside the SubPlan (or InitPlan). The Explicit
Redistribute did not applied correctly inside the apply_motion_mutator function
because by that moment the value segidColIdx had been lost. The problem occured
because previously SubPlans had been mutated inside the ParallelizeSubplan
function, where the SubPlan's plan had been copied via copyObject function.
This function copies the whole plan including the Flow node, which is copied
via _copyFlow function. However the Flow node copying did not include the
copying of segidColIdx Flow field, which is used for valid performance of
Explicit Redistribute Motion.

Therefore, this patch solves the issue by adding segidColIdx to the list of
fields to copy, serialize, deserialize and compare in the _copyFlow, _outFlow,
_readFlow and _equalFlow functions respectively.

bimboterminator1 committed:
Support Explicit Redistribute Motion inside the SubPlans.

When a query had UPDATE/DELETE operation, that required Explicit Redistribute,
inside the SubPlan (or InitPlan), the Ecxplicit Redistribute did not applied
due to an assertion error. The error was thrown from apply_motion_mutator
function and said that the Flow's segidColIdx attribute did not satisfy a
validity check. That happened because, previously SubPlans had been mutated
inside the ParallelizeSubplan function, where the SubPlan's plan had been
copied via copyObject function. This function copies the whole plan including
the Flow node. However the Flow node copying did not include the copying of
segidColIdx Flow field, which is used for correct performing of Explicit
Redistribute Motion.

This patch solves the issue by adding segidColIdx to the list of attributes in
the copyFlow function.

Stolb27 commented on src/test/regress/expected/subselect_gp.out:
```diff
@@ -3112,3 +3112,201 @@ select * from r where b in (select b from s where c=10 order by c limit 2);
  1 | 2 | 3
 (1 row)
 
+-- Test that Explicit Redistribute Motion is applied properly for
```
You should modify `/src/test/regress/expected/subselect_gp_1.out` too. See that file commit history for details. You didn't face CI issues as far as ppc64le tests are muted now.

bimboterminator1 replied:
done

bimboterminator1 committed:
Support Explicit Redistribute Motion inside the SubPlans.

When a query had UPDATE/DELETE operation, that required Explicit Redistribute,
inside the SubPlan (or InitPlan), the Explicit Redistribute did not applied
correctly. On debug build with asserts the assertion error was thrown from
apply_motion_mutator function, which said that the Flow's segidColIdx attribute
did not point to any element of plan's (on top of which Explicit Redistribute
was meant to be applied) target list. This field represents an
index of the gp_segment_id column in the target list, and this column is
subsequently used for Explicit Redistirubte Motion. On release build without
asserts executor could not execute the Explicit Redistribute Motion by
throwing an error related to invalid attribute number or could perform Explicit
Residtibute Motion wrongly because of segidColIdx referencing a wrong column.

The problem occured because previously SubPlans had been mutated inside the
ParallelizeSubplan function, where the SubPlan's plan had been copied via
copyObject function. This function copies the whole plan including the Flow
node. However the Flow node copying did not include the copying of segidColIdx
Flow field, which is used for valid performance of Explicit Redistribute Motion.

This patch solves the issue by adding segidColIdx to the list of attributes in
the copyFlow function.

bimboterminator1 commented:
Have changed the patch description, tests.

Stolb27 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-4149

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/52647

andr-sokolov commented:
@bimboterminator1 , "Residtibute" - typo

bimboterminator1 committed:
Support Explicit Redistribute Motion inside the SubPlans.

When a query had UPDATE/DELETE operation, that required Explicit Redistribute,
inside the SubPlan (or InitPlan), the Explicit Redistribute did not applied
correctly. On debug build with asserts the assertion error was thrown from
apply_motion_mutator function, which said that the Flow's segidColIdx attribute
did not point to any element of plan's (on top of which Explicit Redistribute
was meant to be applied) target list. This field represents an
index of the gp_segment_id column in the target list, and this column is
subsequently used for Explicit Redistribute Motion. On release build without
asserts executor could not execute the Explicit Redistribute Motion by
throwing an error related to invalid attribute number or could perform Explicit
Redistribute Motion wrongly because of segidColIdx referencing a wrong column.

The problem occured because previously SubPlans had been mutated inside the
ParallelizeSubplan function, where the SubPlan's plan had been copied via
copyObject function. This function copies the whole plan including the Flow
node. However the Flow node copying did not include the copying of segidColIdx
Flow field, which is used for valid performance of Explicit Redistribute Motion.

This patch solves the issue by adding segidColIdx to the list of attributes in
the copyFlow function.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/52668

andr-sokolov commented:
I suggest to fix the _outFlow function too

bimboterminator1 committed:
Fix Flow structure copying, comparison, serialization and deserialization.

Flow structure has a segidColIdx field, which referencing a gp_segment_id
column in plan's targetList when the Explicit Redistribute Motion is requested.
There was a problem, that _copyFlow, _equalFlow, _outFlow and _readFlow
functions did not handled the segidColIdx field. Therefore, the Flow node
could not be serialized and deserialized, or copied corretly.

The problem manifested itself when a query had UPDATE/DELETE operation, that
required Explicit Redistribute, inside the SubPlan (or InitPlan). The Explicit
Redistribute did not applied correctly inside the apply_motion_mutator function
because by that moment the value segidColIdx had been lost. The problem occured
because previously SubPlans had been mutated inside the ParallelizeSubplan
function, where the SubPlan's plan had been copied via copyObject function.
This function copies the whole plan including the Flow node, which is copied
via _copyFlow function. However the Flow node copying did not include the
copying of segidColIdx Flow field, which is used for valid performance of
Explicit Redistribute Motion.

Therefore, this patch solves the issue by adding segidColIdx to the list of
fields to copy, serialize, deserialize and compare in the _copyFlow, _outFlow,
_readFlow and _equalFlow functions respectively.

bimboterminator1 committed:
Fix Flow structure copying, comparison, serialization and deserialization.

Flow structure has a segidColIdx field, which referencing a gp_segment_id
column in plan's targetList when the Explicit Redistribute Motion is requested.
There was a problem, that _copyFlow, _equalFlow, _outFlow and _readFlow
functions did not handled the segidColIdx field. Therefore, the Flow node
could not be serialized and deserialized, or copied correctly.

The problem manifested itself when a query had UPDATE/DELETE operation, that
required Explicit Redistribute, inside the SubPlan (or InitPlan). The Explicit
Redistribute did not applied correctly inside the apply_motion_mutator function
because by that moment the value segidColIdx had been lost. The problem occured
because previously SubPlans had been mutated inside the ParallelizeSubplan
function, where the SubPlan's plan had been copied via copyObject function.
This function copies the whole plan including the Flow node, which is copied
via _copyFlow function. However the Flow node copying did not include the
copying of segidColIdx Flow field, which is used for valid performance of
Explicit Redistribute Motion.

Therefore, this patch solves the issue by adding segidColIdx to the list of
fields to copy, serialize, deserialize and compare in the _copyFlow, _outFlow,
_readFlow and _equalFlow functions respectively.

bimboterminator1 commented:
Have changed the patch description, the final description is in the latest commit.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/52699

bimboterminator1 committed:
Fix Flow structure copying, comparison, serialization and deserialization.

Flow structure has a segidColIdx field, which referencing a gp_segment_id
column in plan's targetList when the Explicit Redistribute Motion is requested.
There was a problem, that _copyFlow, _equalFlow, _outFlow and _readFlow
functions did not handle the segidColIdx field. Therefore, the Flow node
could not be serialized and deserialized, or copied correctly.

The problem manifested itself when a query had UPDATE/DELETE operation, that
required Explicit Redistribute, inside the SubPlan (or InitPlan). The Explicit
Redistribute did not applied correctly inside the apply_motion_mutator function
because by that moment the value segidColIdx had been lost. The problem occured
because previously SubPlans had been mutated inside the ParallelizeSubplan
function, where the SubPlan's plan had been copied via copyObject function.
This function copies the whole plan including the Flow node, which is copied
via _copyFlow function. However the Flow node copying did not include the
copying of segidColIdx Flow field, which is used for valid performance of
Explicit Redistribute Motion.

Therefore, this patch solves the issue by adding segidColIdx to the list of
fields to copy, serialize, deserialize and compare in the _copyFlow, _outFlow,
_readFlow and _equalFlow functions respectively.

bimboterminator1 committed:
Correct binary format

bimboterminator1 committed:
Rerun tests
